### ISSUE TITLE
Vital Sign Detection + RVF Container: Breathing & Heart Rate from WiFi CSI

### ISSUE THREAD

**Issue by ruvnet** [ruvnet/wifi-densepose#45]
## Summary

This issue tracks the implementation of **contactless vital sign detection** (breathing rate and heart rate) from WiFi Channel State Information (CSI) and the **RVF container format** for single-file trained model deployment. These capabilities are defined in [ADR-021](docs/adr/ADR-021-vital-sign-detection-rvdna-pipeline.md) and [ADR-023](docs/adr/ADR-023-trained-densepose-model-ruvector-pipeline.md).

WiFi CSI captures fine-grained multipath propagation changes caused by physiological movements -- chest displacement from respiration (1-5 mm amplitude, 0.1-0.5 Hz) and body surface displacement from cardiac activity (0.1-0.5 mm, 0.8-2.0 Hz). This extends the project from occupancy sensing into **health monitoring**, enabling contactless respiratory rate and heart rate estimation for eldercare, sleep monitoring, disaster survivor detection, and clinical triage.

### Vital Sign Detection

| Capability | Range | Method |
|------------|-------|--------|
| Breathing Rate | 6-30 BPM (0.1-0.5 Hz) | Bandpass filter + FFT peak detection on CSI amplitude |
| Heart Rate | 40-120 BPM (0.8-2.0 Hz) | Bandpass filter + FFT peak detection on CSI phase |
| Sampling Rate | 20 Hz | ESP32 CSI frame rate |
| Confidence Scoring | 0.0-1.0 | Spectral coherence + SNR-based quality assessment |
| Motion Artifact Rejection | Automatic | DVS event-based gating during large body movements |
| Multi-subcarrier Fusion | Attention-weighted | GNN-learned subcarrier correlation topology |
| Anomaly Detection | z-score, CUSUM, EMA | Biomarker-stream pattern from rvdna DNA example |

### RVF Container Format

| Property | Detail |
|----------|--------|
| Format | Segment-based binary, magic `0x52564653`, 64-byte headers |
| Contents | Model weights + HNSW index + metadata + WASM runtime |
| Progressive Loading | Layer A <5ms, Layer B 100ms-1s, Layer C full graph |
| Signing | Ed25519 training proofs for verifiable provenance |
| Quantization | Temperature-tiered f32/f16/u8 via `rvf-quant` |
| CLI Integration | `--save-rvf` and `--load-rvf` flags |

## API Endpoints

| Endpoint | Method | Protocol | Description |
|----------|--------|----------|-------------|
| `/api/v1/vital-signs` | GET | REST | Latest HR, RR, confidence, and timestamp |
| `/ws/sensing` | - | WebSocket | Real-time stream; each frame includes `vital_signs` object |
| `/api/v1/vital-signs/history` | GET | REST | Historical vital sign readings (tiered storage) |
| `/api/v1/vital-signs/config` | GET/PUT | REST | Vital sign extraction configuration |

### Example Response (`GET /api/v1/vital-signs`)

```json
{
  "breathing_rate_bpm": 14.2,
  "heart_rate_bpm": 72.5,
  "breathing_confidence": 0.87,
  "heart_rate_confidence": 0.63,
  "motion_artifact": false,
  "timestamp_ms": 1709136000000,
  "source": "esp32"
}
```

## Usage Examples

```bash
# Build the sensing server
cd rust-port/wifi-densepose-rs
cargo build --release -p wifi-densepose-sensing-server

# Run with simulation (no hardware needed)
./target/release/sensing-server --source simulate --ui-path ../../ui

# Run with ESP32 hardware
./target/release/sensing-server --source esp32 --ui-path ../../ui

# Query vital signs
curl http://localhost:8080/api/v1/vital-signs

# Save trained model as RVF container
./target/release/sensing-server --source simulate --save-rvf model.rvf

# Load model from RVF container
./target/release/sensing-server --source esp32 --load-rvf model.rvf

# Pre-train on MM-Fi public dataset
cargo run -p wifi-densepose-train -- --dataset mmfi --epochs 100 --save-rvf pretrained.rvf

# Fine-tune with local ESP32 data
cargo run -p wifi-densepose-train -- --dataset local --fine-tune \
  --base-model pretrained.rvf --epochs 20 --save-rvf finetuned.rvf
```

## Architecture

```
ESP32 CSI (UDP:5005)
    |
    v
+--------------------------------------------------------------+
|  wifi-densepose-signal (existing)                            |
|  Phase sanitization, Hampel filter, subcarrier selection     |
+-------------------------------+------------------------------+
                                |
                                v
+--------------------------------------------------------------+
|  wifi-densepose-vitals (new crate, ADR-021)                  |
|                                                              |
|  CsiVitalPreprocessor (ruvector-nervous-system)              |
|    PredictiveLayer: suppress static CSI (90-99% reduction)   |
|    EventRingBuffer: streaming frame ingestion                |
|                |                                             |
|                v                                             |
|  SubcarrierWeighter (ruvector-attention + ruvector-gnn)      |
|    Attention-weighted selection of vital-sign-sensitive       |
|    subcarriers using GNN correlation graph                   |
|                |                                             |
|                v                                             |
|  VitalSignExtractor                                          |
|    BreathingExtractor: bandpass 0.1-0.5 Hz + FFT             |
|    HeartRateExtractor: bandpass 0.8-2.0 Hz + FFT             |
|    MotionArtifactRejector: DVS event gating                  |
|                |                                             |
|                v                                             |
|  VitalCoherenceGate (ruQu three-filter pipeline)             |
|    Structural / Shift / Evidence filters                     |
|    DriftDetector for baseline tracking                       |
|                |                                             |
|                v                                             |
|  VitalSignStore (ruvector-temporal-tensor)                   |
|    Hot (real-time, 8-bit) / Warm (5-bit) / Cold (3-bit)     |
+-------------------------------+------------------------------+
                                |
                    +-----------+-----------+
                    |                       |
                    v                       v
          REST API                WebSocket /ws/sensing
      /api/v1/vital-signs        (vital_signs field)
```

```
+--------------------------------------------------------------+
|  Trained DensePose Pipeline (ADR-023)                        |
|                                                              |
|  RuVector Signal Preprocessing Layer (5 crates)              |
|    attn-mincut -> mincut -> attention -> solver -> temporal   |
|                |                                             |
|                v                                             |
|  Neural Network (6 crates)                                   |
|    ModalityTranslator (graph-transformer)                    |
|    GNN Body Graph (ruvector-gnn)                             |
|    DensePose Head (17 keypoints + 25 parts + 48 UV)          |
|                |                                             |
|                v                                             |
|  RVF Container (.rvf file)                                   |
|    Weights + HNSW index + WASM runtime + Ed25519 proofs      |
|    Progressive loading: <5ms startup                         |
+--------------------------------------------------------------+
```

## Performance Targets

| Metric | Target | Notes |
|--------|--------|-------|
| Vital sign extraction latency | <50ms per frame | At 20 Hz ESP32 frame rate |
| Breathing rate accuracy | +/- 1 BPM | Validated against reference sensor |
| Heart rate accuracy | +/- 5 BPM | WiFi cardiac signal is weaker than respiration |
| RVF load time (Layer A) | <5ms | Entry points only, instant startup |
| RVF full load time | <5s | Complete model with HNSW index |
| DensePose inference | <50ms per frame | With ruvector-sparse-inference |
| SONA adaptation | <0.05ms | Micro-LoRA parameter update |

## Training Data Strategy

Three-tier approach:

1. **Pre-train** on [MM-Fi](https://github.com/ybhbingo/MMFi_dataset) and [Wi-Pose](https://github.com/aiot-lab/Wi-Pose) public datasets for cross-environment generalization
2. **Fine-tune** with ESP32 self-collected data using camera-based teacher model for pseudo-labels
3. **SONA runtime adaptation** via micro-LoRA + EWC++ for continuous on-device learning without catastrophic forgetting

## RuVector Crates Used (11)

| Crate | Description |
|-------|-------------|
| `ruvector-core` | VectorDB, HNSW index, SIMD distance |
| `ruvector-attention` | Scaled dot-product, MoE, sparse attention |
| `ruvector-gnn` | Graph neural network, graph attention |
| `ruvector-nervous-system` | PredictiveLayer, OscillatoryRouter, EventBus |
| `ruvector-coherence` | Spectral coherence, signal quality |
| `ruvector-temporal-tensor` | Tiered temporal compression |
| `ruvector-mincut` | Dynamic min-cut for multi-person |
| `ruvector-attn-mincut` | Attention-gated min-cut |
| `ruvector-solver` | Sparse Neumann solver |
| `ruvector-graph-transformer` | Proof-gated graph transformer |
| `ruvector-sparse-inference` | PowerInfer-style sparse execution |

## Related ADRs

- [ADR-021: Vital Sign Detection via rvdna Signal Processing Pipeline](docs/adr/ADR-021-vital-sign-detection-rvdna-pipeline.md)
- [ADR-023: Trained DensePose Model with RuVector Signal Intelligence Pipeline](docs/adr/ADR-023-trained-densepose-model-ruvector-pipeline.md)
- [ADR-003: RVF Cognitive Containers](docs/adr/ADR-003-rvf-cognitive-containers.md)
- [ADR-005: SONA Self-Learning](docs/adr/ADR-005-sona-self-learning.md)
- [ADR-014: SOTA Signal Processing](docs/adr/ADR-014-sota-signal-processing.md)
- [ADR-015: Public Dataset Strategy](docs/adr/ADR-015-public-dataset-strategy.md)
- [ADR-020: Rust AI Migration](docs/adr/ADR-020-rust-ruvector-ai-model-migration.md)

## Implementation Checklist

- [ ] Create `wifi-densepose-vitals` crate with module structure
- [ ] Implement `CsiVitalPreprocessor` using `ruvector-nervous-system::PredictiveLayer`
- [ ] Implement `BreathingExtractor` (0.1-0.5 Hz bandpass + FFT)
- [ ] Implement `HeartRateExtractor` (0.8-2.0 Hz bandpass + FFT)
- [ ] Implement `SubcarrierWeighter` using attention + GNN
- [ ] Implement `MotionArtifactRejector` with DVS event gating
- [ ] Implement `VitalCoherenceGate` using ruQu three-filter pipeline
- [ ] Implement `VitalSignStore` with tiered temporal compression
- [ ] Add `GET /api/v1/vital-signs` REST endpoint to sensing server
- [ ] Add `vital_signs` field to WebSocket `/ws/sensing` frames
- [ ] Implement `--save-rvf` and `--load-rvf` CLI flags
- [ ] Integrate MM-Fi dataset loader for pre-training
- [ ] Implement SONA runtime adaptation loop
- [ ] Add unit tests for each vital sign extraction stage
- [ ] Add integration test with simulated CSI containing known breathing/HR signals
- [ ] Update UI to display breathing rate and heart rate

**Comment by ruvnet** [comment]
## Implementation Complete

All vital sign detection and RVF container work is now merged. Here's the final status:

### Vital Signs (ADR-021)
- Pure-Rust FFT (radix-2 DIT) + FIR bandpass filter
- Breathing: 0.1–0.5 Hz (6–30 BPM), Heartbeat: 0.8–2.0 Hz (48–120 BPM)
- Rolling buffers, signal quality estimation, confidence scoring
- REST `/api/v1/vital-signs` + WebSocket `vital_signs` field

### RVF Container
- 12 segment types (Vec, Manifest, Quant, Meta, Witness, Profile, Index, Overlay, Crypto, WASM, Dashboard, AggregateWeights)
- CRC32 integrity, 64-byte alignment, progressive 3-layer loading
- `RvfModelBuilder` high-level API for model packaging

### Full DensePose Pipeline (ADR-023)
Beyond the original scope of this issue, the full training pipeline is now implemented:
- **Dataset loaders**: MM-Fi (.npy) + Wi-Pose (.mat) with subcarrier resampling
- **Graph transformer**: COCO BodyGraph, AntennaGraph, multi-head CrossAttention, GCN
- **Training loop**: 6-term loss, SGD+momentum, cosine scheduler, PCK/OKS metrics
- **SONA adaptation**: LoRA (rank-4) + EWC++ Fisher regularization
- **Sparse inference**: Hot/cold neuron profiling, INT8/FP16 quantization
- **Server integration**: `--model`, `--progressive`, 4 new REST endpoints

### Numbers
| Metric | Value |
|--------|-------|
| Total tests | 229 (all passing) |
| Throughput | 9,520 frames/sec (105μs/frame) |
| Real-time factor | 476x at 20 Hz |
| Total Rust lines | 7,832 |
| External ML deps | 0 |

Commits: `1192de95` (vital signs + RVF base) and `fc409dfd` (full pipeline).

— Ruflo AI

**Comment by ruvnet** [comment]
## Docker + RVF Available

Vital sign detection is included in both Docker images:

```bash
docker pull ruvnet/wifi-densepose:latest
docker run -p 3000:3000 ruvnet/wifi-densepose:latest

# Check vital signs
curl http://localhost:3000/api/v1/vital-signs
```

RVF model packages include vital sign config (breathing 0.1-0.5 Hz, heartbeat 0.8-2.0 Hz):
```bash
docker run --rm -v $(pwd):/out ruvnet/wifi-densepose:latest --export-rvf /out/wifi-densepose-v1.rvf
```

Benchmark: 11,665 fps on release build.
